### PR TITLE
[TPU VM] Do not clean up again if the cluster is already removed

### PR DESCRIPTION
**Pull request by Michaelvll**
Describe the changes in this PR:

This is a follow up fix for #1555, i.e. we don't want to do the post teardown clean up again, if the cluster is already terminated, as we will always post teardown cleanup before the cluster is removed from the cluster table.

Tested (run the relevant ones):

- [x] Any manual or new tests for this the PR
  - `sky launch -c tpu-vm --use-spot test-tpuvm.yaml`; delete the tpu vm from the console; `sky down tpu-vm`. With the current master branch, it will show the following error:
  ```
    File "/Users/zhwu/Library/CloudStorage/OneDrive-Personal/AResource/PhD/Research/sky-computing/code/sky-experiment-dev/sky/backends/cloud_vm_ray_backend.py", line 2473, in _teardown
      success = self.teardown_no_lock(handle, terminate, purge)
    File "/Users/zhwu/Library/CloudStorage/OneDrive-Personal/AResource/PhD/Research/sky-computing/code/sky-experiment-dev/sky/backends/cloud_vm_ray_backend.py", line 2786, in teardown_no_lock
      return self.post_teardown_cleanup(handle, terminate, purge)
    File "/Users/zhwu/Library/CloudStorage/OneDrive-Personal/AResource/PhD/Research/sky-computing/code/sky-experiment-dev/sky/backends/cloud_vm_ray_backend.py", line 2834, in post_teardown_cleanup
      config = common_utils.read_yaml(handle.cluster_yaml)
    File "/Users/zhwu/Library/CloudStorage/OneDrive-Personal/AResource/PhD/Research/sky-computing/code/sky-experiment-dev/sky/utils/common_utils.py", line 166, in read_yaml
      with open(path, 'r') as f:
  FileNotFoundError: [Errno 2] No such file or directory: '/Users/zhwu/.sky/generated/tpu-vm.yml'
  ```
- [x] `tests/run_smoke_tests.sh test_tpu`
- [x] `tests/run_smoke_tests.sh test_tpu_vm`
- [x] `tests/run_smoke_tests.sh test_tpu_vm_pod`

### CODE REVIEW

File: sky/backends/cloud_vm_ray_backend.py [hunk 1]
```diff
@@ -2782,7 +2782,10 @@ def teardown_no_lock(self,
                         stderr=stderr))
                 return False
 
-        if post_teardown_cleanup:
+        # No need to clean up when the cluster is already terminated
+        # (prev_status), as the clean up will always be done after the cluster
+        # is removed from the status table.
```
Comment (infwinston) [inline sky/backends/cloud_vm_ray_backend.py:2787]: small nit:

```suggestion
        # No need to clean up if the cluster is already terminated
        # (i.e., prev_status is None), as the cleanup has already been done
        # if the cluster is removed from the status table.
```

**Review by infwinston — COMMENTED** [review]
Thanks a lot for fixing this @Michaelvll ! just left some small nits.

another nit: do we want to refactor the below place with `remove_file_if_exists`?
https://github.com/skypilot-org/skypilot/blob/1bdeadca142af57f56621456bde590210803e2e9/sky/backends/backend_utils.py#L661-L662

One question I don't fully understand is why the below line was not triggered if the cluster yaml was already gone? do you know why?
https://github.com/skypilot-org/skypilot/blob/1bdeadca142af57f56621456bde590210803e2e9/sky/backends/cloud_vm_ray_backend.py#L2707

File: sky/utils/common_utils.py [hunk 2]
```diff
@@ -336,3 +336,15 @@ def remove_color(s: str):
...
+
+def remove_file_if_exists(path: str):
+    """Delete a file if it exists.
+
+    Args:
+        path: The path to the file.
+    """
+    try:
+        os.remove(path)
+    except FileNotFoundError:
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/utils/common_utils.py:349]: Good point! Done. Thanks!

**Comment by Michaelvll** [comment]
> Another question I don't fully understand is why the below line was not triggered if the cluster yaml was already gone? do you know why?

It is a great question. Actually, we should skip not only the post_teardown but also the whole termination process if the cluster does not exist in our cluster table.

**Review by infwinston — APPROVED** [review]
Great catch and thanks again for the fix! the changes look good to me now.

**Pull request merged**